### DISCUSSION TITLE
ViperTypeError: div/mod not implemented for uint

### DISCUSSION THREAD

**Discussion by samneggs** [micropython/micropython#14991] [RP2040 / Pico] [answered]
Just to verify - even though the RP2040 has hardware div/mod for both int and uint, only int works with Viper?
Sam

**Comment by bixb922** [comment, marked as answer]
That is correct, no div/mod with Viper uint

Here is a detailed description of Viper, this restriction (and others) are mentioned there:

https://github.com/micropython/micropython/wiki/Improving-performance-with-Viper-code

**Comment by samneggs** [comment]
All good stuff.
So uint works with addition, subtraction, multiplication but not division.
Seems a bit arbitrary but whatever.
Thanks!

**Reply by GitHubsSilverBullet** [reply]
One could use something like:
```py
@micropython.viper
def udiv(a: uint, b: uint) -> uint:
    return uint(int(a) // b)
```
Note: Only the first parameter must be converted to int before division.
Note 2: Has no error checking. As soon as the MSB is set, all bets are off.

**Comment by samneggs** [comment]
Here is a solution for uint divides that's 1.5x times faster than // and you get a remainder.
```
@micropython.viper
def bench():
    dividend = ptr32(0xd000_0060)
    divisor  = ptr32(0xd000_0064)
    quotient = ptr32(0xd000_0070)
    ticks = ticks_ms()
    j1 :int = 0
    j2 :int = 0
    dummy : int = 0
    for i in range(1<<20):
        dividend[0] = i
        divisor[0] = 3                # 386 ms
        dummy : int = 0
        dummy : int = 0
        dummy : int = 0
        j2 = quotient[0]
        r  = quotient[1]
        #j1 = i//3                    #604 ms
    print(ticks_diff(ticks_ms(),ticks))
    print(i,j1,j2)

**Reply by rkompass** [reply]
Which works on RP2. Nice.
Do you have to divide by a constant value of 3 always? Then division by multiplication may be an idea.

**Reply by samneggs** [reply]
It can be a variable, this was just a quick demo.

**Comment by peterhinch** [comment]
I must confess I can't understand how this works. Here is my attempt. Runs in ~9μs on the Pico at standard 125MHz clock.
```py
from time import ticks_us, ticks_diff
@micropython.viper
def div(n : int, d: int) -> int:  # n: numerator d: denominator
    t = ticks_us()
    m : int = 0
    q : int = 0  # Quotient
    while n > d:
        d <<= 1
        m += 1
    m -= 1
    d >>= 1
    while m > -1:
        q <<= 1
        if n >= d:
            n -= d
            q |= 1
        d >>= 1
        m -= 1
    #print(q, n)  # Remainder in n
    dt = ticks_diff(ticks_us(), t)
    print(f"Runtime {dt}μs")
    return q
```
Example session on Pico at standard clock:
```py
>>> div(1_234_567, 739)
Runtime 9μs
1670
>>> 1_234_567//739
1670
```
The remainder is available in `n`.

**Reply by GitHubsSilverBullet** [reply]
It uses the hardware divider directly. Takes eight CPU cycles per operation.
The three `dummy: int = 0` are only there to provide the eight cycle wait time until the divider finishes.

**Comment by peterhinch** [comment]
Or this variant where an array is populated with quotient and remainder:
```py
@micropython.viper
def div(n : int, d: int, ar: ptr32) -> int:
    m : int = 0
    q : int = 0
    while n > d:
        d <<= 1
        m += 1
    m -= 1
    d >>= 1
    while m >= 0:
        q <<= 1
        if n >= d:
            n -= d
            q |= 1
        d >>= 1
        m -= 1
    ar[0] = q
    ar[1] = n
    return q
```
Session:
```py
from array import array
ar = array("I", (0,0))
div(9_876_543, 847, ar)
print(ar)
divmod(9_876_543, 847)
11660
array('I', [11660, 523])
(11660, 523)
```

**Comment by samneggs** [comment]
@peterhinch your discrete bit shift method was indeed faster (initially)  than the built-in // and the RP2040 hardware with pointers. 
I ran loops of 1,10,100,1000,10000 of 1_234_567/739 .There seems to be some overhead but each method scaled linearly. 
So... application matters and always benchmark.   


![image](https://github.com/micropython/micropython/assets/59456193/f6561ea8-6348-44dd-8952-96dcf5f80bcc)

**Reply by GitHubsSilverBullet** [reply]
> I ran loops of 1,10,100,1000,10000 of 1_234_567/739 .There seems to be some overhead but each method scaled linearly.

If one divides by 2,4,8,… then, and only then, an assembler shift instruction is faster than using the hardware divider.
If you measured that manually dividing by 739 is faster than the (fixed) 8 cycle hardware divider, your measurement setup is flawed.
Just sayin'…
```python
#!micropython
# -*- coding: UTF-8 -*-
# vim:fileencoding=UTF-8:ts=4

from time import ticks_us, ticks_diff
from machine import freq as cpufreq

SIO_BASE        = const(0xd0000000)
DIV_UDIVIDEND   = const(0x060)
DIV_UDIVISOR    = const(0x064)
DIV_QUOTIENT    = const(0x070)
DIV_REMAINDER   = const(0x074)

LOOPS = const(1_000_000)

@micropython.viper
def bench():
    sio = ptr32(SIO_BASE)
    dividend: int = 1_234_567
    divisor: int = 739
    quotient: int
    remainder: int
    loopcnt: int

    loopcnt = LOOPS
    t0 = ticks_us()
    while loopcnt:
        loopcnt -= 1
    t1 = ticks_us()
    td_empty = ticks_diff(t1, t0)

    loopcnt = LOOPS
    t0 = ticks_us()
    while loopcnt:
        sio[DIV_UDIVIDEND>>2] = dividend
        sio[DIV_UDIVISOR>>2] = divisor
        loopcnt -= 1
        quotient = sio[DIV_QUOTIENT>>2]
        remainder = sio[DIV_REMAINDER>>2]
    t1 = ticks_us()
    td_µs = ticks_diff(t1,t0) - td_empty
    t_ns = td_µs * 1000.0 / float(LOOPS)

    print(f'{t_ns:.1f} ns/div   (expected {8e9/cpufreq()} ns @{cpufreq()/1e6}MHz)')
    print(quotient, remainder)

bench()
print(divmod(1_234_567, 739))

# => 64.0 ns/div   (expected 64.0 ns @125.0MHz)
# => 1670 437
# => (1670, 437)
```

**Reply by samneggs** [reply]
@GitHubsSilverBullet my point was that you need to account for the overhead and the number of loops. Try only one divide (no loop) and include sio = ptr32(SIO_BASE) in the total time. 
Dividing by factor of two using bit shifting, fastest , agree.
Dividing by other integers using hardware divide and viper pointers with many loops, second fastest, agree.
Dividing by other integers using hardware divide and viper pointers only once, may be slower than discreet math divide.

btw, I like how you used this as the delay:
```
loopcnt -= 1
